vowpalwabbit/learner.h

```diff
@@ -401,6 +411,22 @@ struct learner
     save_load_fd.base = learn_fd.base;
   }
 
+  // called when metrics is enabled.  Autorecursive.
+  void persist_metrics(std::vector<std::tuple<std::string, size_t>>& metrics_list)
+  {
+    persist_metrics_fd.save_metric_f(persist_metrics_fd.data, metrics_list);
+    if (persist_metrics_fd.base) persist_metrics_fd.base->persist_metrics(metrics_list);
+  }
+  void set_persist_metrics(void (*f)(T&, std::vector<std::tuple<std::string, size_t>>&))
```

Comment:
So does this mean that the only metrics which can be reported are `size_t`?

Reply:
for now, yes. We eventually want to support floats too.
It's not that bad to restrict, to avoid reductions holding into too much state during the runtime.